src/line_launcher/mod.rs

```diff
@@ -112,149 +146,292 @@ impl LineLauncher {
+                    Some(0.5),
+                    listen_for_control_changes(
+                        midi_message_bus.add_rx(),
+                        ahead_or_behind_the_beat_ratio_slider,
+                    ),
+                ),
+                None => (None, {
+                    let (_sender, receiver) = mpsc::channel();
+                    receiver
+                }),
+            };
         if let Some(midi_messages) = midi_messages {
             thread::spawn(move || {
                 for midi_message in midi_messages.iter() {
                     midi_message_bus.broadcast(midi_message);
                 }
             });
         }
-        for message in
-            get_combined_message_receiver(beat_message_receiver, duration_ratio_receiver).iter()
+        for message in get_combined_message_receiver(
+            beat_message_receiver,
+            duration_ratio_receiver,
+            ahead_or_behind_the_beat_ratio_receiver,
+            fire_note_on_receiver,
+            fire_note_off_receiver,
+        )
+        .iter()
         {
             match message {
                 CombinedMessage::BeatMessage(beat_message) => {
+                    debug!(
+                        "beat message: {:?}, now: {:?}",
+                        beat_message,
+                        SystemTime::now()
+                    );
                     duration_between_sixteenth_notes =
                         duration_between_sixteenth_notes.process_beat_message(&beat_message);
-                    if beat_message.is_beginning_of_measure() {
+                    if (!progression_state.has_started() && beat_message.is_beginning_of_measure())
+                        || beat_message.is_next_beginning_of_measure()
+                    {
                         progression_state.tick_measure();
                     }
-                    let mut state = state_mutex.lock().unwrap();
-                    *state = match *state {
-                        PlayingState::NotPlaying if beat_message.is_beginning_of_measure() => {
-                            *state = PlayingState::Playing {
+                    match playing_state {
+                        PlayingState::NotPlaying
+                            if beat_message.is_next_beginning_of_measure()
+                                && matches!(
+                                    duration_between_sixteenth_notes.get_duration(),
+                                    Some(_)
+                                ) =>
+                        {
+                            playing_state = PlayingState::Playing {
                                 line_index: rand::thread_rng().gen_range(0..self.lines.len()),
                                 next_note_index: 0,
                                 pitch_offset: progression_state.current_chord().pitch.index(),
-                                has_fired_previous_note_off: true,
+                                next_note_off_index: 0,
```

Comment:
I'm not 100% sure if this was necessary but changed `has_fired_previous_note_off` to the more explicit `next_note_off_index` for tracking which note off's we've fired already